src/parsing/binast-deserializer-inl.h

```diff
@@ -268,21 +270,28 @@ inline BinAstDeserializer::DeserializeResult<Variable*> BinAstDeserializer::Dese
   }
 
   auto variable_result = variables_by_id_.find(variable_reference.value);
-  DCHECK(variable_result != variables_by_id_.end());
-  Variable* variable = variable_result->second;
+  DCHECK(variable_result != variables_by_id_.end() || scope);
 
-  return {variable, offset};
+  if (variable_result != variables_by_id_.end()) {
+    return {variable_result->second, offset};
+  } else {
+    return {
+      DeserializeScopeVariable(serialized_binast, variable_reference.value, scope).value,
+      offset
+    };
+  }
```

Comment:
What case are we handling here? When wouldn't the variable be found in variables_by_id_? Might be good to add a comment explaining.